Fabric8 deletion behaviour fix

stanlyDoge opened the pull request:
Signed-off-by: Stanislav Knot <sknot@redhat.com>

### Type of change
- Bugfix

### Description
Fixes https://github.com/strimzi/strimzi-kafka-operator/issues/2223

After https://github.com/fabric8io/kubernetes-client/pull/1807 was merged in fabric8 I think this bug should be fixed.
//edit: using 4.6.4 did not seem to help, using suggested `cascading(true)` where possible.

### Checklist
- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

stanlyDoge commented:
@strimzi-ci run tests

tombentley commented on topic-operator/src/test/java/io/strimzi/operator/topic/TopicOperatorBaseIT.java:
```diff
@@ -200,7 +200,7 @@ public void teardown() throws InterruptedException, TimeoutException, ExecutionE
                 for (KafkaTopic item : items) {
                     LOGGER.info("Deleting {} from Kube", item.getMetadata().getName());
-                    operation().inNamespace(NAMESPACE).withName(item.getMetadata().getName()).cascading(true).delete();
+                    operation().inNamespace(NAMESPACE).withName(item.getMetadata().getName()).delete();
```
I think there's another place in `TopicOperatorBaseIT` where we do a cascading delete (see [85eaa7fe](https://github.com/strimzi/strimzi-kafka-operator/commit/85eaa7fea4fee32ea0582dd918a4e8f1fe3d87da#diff-25da95bbedc25f0bd60e013de43219c7R600))

stanlyDoge replied:
I reverted just https://github.com/strimzi/strimzi-kafka-operator/pull/2207/files for now... If the test results will be ok, I will search for other occurrences. However I did not know there any other, so thank you for pointing it out.

stanlyDoge commented:
@strimzi-ci run tests

stanlyDoge commented:
@strimzi-ci run tests

stanlyDoge commented:
@strimzi-ci run tests

stanlyDoge commented:
After a few tests runs it seems there is still an issue with deleting resources, so I reverted my reversion and added `cascading(true)` to all occurrences of `delete(resource)`.

tombentley commented on mockkube/src/test/java/io/strimzi/test/io/strimzi/test/mockkube/MockKubeTest.java:
```diff
@@ -256,7 +256,7 @@ public void podNameScopedCreateListGetDelete(Class<RT> cls,
         assertThat(gotResource, is(nullValue()));
 
-        assertThat(mixedOp.apply(client).withName(pod.getMetadata().getName()).delete(), is(false));
+        assertThat(mixedOp.apply(client).withName(pod.getMetadata().getName()).cascading(true).delete(), is(false));
```
I don't think the MockKube properly supports cascading deletion (i.e. I suspect it behaves the same for a cascading as a non-cascading delete), but if the tests pass I guess it doesn't matter.

stanlyDoge replied:
Yes, the test pass.

scholzj commented:
TBH, I'm a bit unsure abotu this.
* It seems ot be adding it to delete operations where no cascading should be happening
* It seems to add it only to tests. Do we have no delete operations what so ever in our regular code? Is it that we delete everything with GC and don't use it in the actual applications?

The STs seem to have passed, so I'm fine with this as it does not seem to do any damage. But I'm not sure I understand this.

stanlyDoge commented:
I remember this "not so clear" modification was used in some topic ST. I tried to remove `cascading(true)` there and the test become flaky. Adding `cascading(true)` where applicable should prevent tests from being flaky in the matter of deleting resources.
